clang/lib/3C/ArrayBoundsInferenceConsumer.cpp

```diff
@@ -703,12 +725,37 @@ class ComparisionVisitor : public RecursiveASTVisitor<ComparisionVisitor> {
     if (BO->getOpcode() == BO_LT || BO->getOpcode() == BO_GE) {
       Expr *LHS = BO->getLHS()->IgnoreParenCasts();
       Expr *RHS = BO->getRHS()->IgnoreParenCasts();
-      auto LHSCVars = CR->getExprConstraintVars(LHS);
-      auto RHSCVars = CR->getExprConstraintVars(RHS);
+      BoundsKey LKey, RKey;
+      if (tryGetValidBoundsKey(LHS, LKey, I, C) &&
+          tryGetValidBoundsKey(RHS, RKey, I, C)) {
+        // If this the left hand side of a < comparison and
+        // the LHS is the index used in array indexing operation?
+        // Then add the RHS to the possible bounds key.
+        bool IsRKeyBound = (LKey == IndxBKey);
+        if (BO->getOpcode() == BO_GE) {
+          // If we have: x >= y, then this has to be an IfStmt to
+          // consider Y as upper bound.
+          // Why? This is to distinguish between following cases:
+          // In the following case, we should not
+          // consider y as the bound.
+          // for (i=n-1; i >= y; i--) {
+          //      arr[i] = ..
+          // }
+          // Where as the following is a valid case. MAX_LEN is the bound.
+          // if (i >= MAX_LEN) {
+          //     return -1;
+          //  }
+          //  arr[i] = ..
+          IsRKeyBound &= (CurrStmt != nullptr && isa<IfStmt>(CurrStmt));
+        }
+
+        if (IsRKeyBound)
+          PB.insert(RKey);
+      }
 
-      if (!CR->containsValidCons(LHSCVars) &&
+      /*if (!CR->containsValidCons(LHSCVars) &&
```

Comment:
Here and above: Delete the code that's commented out, or leave an explanation nearby for why you want to leave it here.